### PR TITLE
Make help target and VCS/Verilator Cleanup

### PR DESCRIPTION
**Pull request by abejgonzalez**
**Related issue**: #562

**Type of change**: other enhancement

**Impact**: other

**Release Notes**
This is a split of PR #562 with Makefile change minus the Multi-threading support. It includes changes to VCS so that it aligns with the same changes done to Verilator.

### CODE REVIEW

File: docs/Simulation/Software-RTL-Simulation.rst [hunk 1]
```diff
@@ -132,29 +141,6 @@ All ``make`` targets that can be applied to the default example, can also be app
 Finally, in the ``generated-src/<...>-<package>-<config>/`` directory resides all of the collateral and Verilog source files for the build/simulation.
 Specifically, the SoC top-level (``TOP``) Verilog file is denoted with ``*.top.v`` while the ``TestHarness`` file is denoted with ``*.harness.v``.
 
-Fast Memory Loading
```
Comment (jerryz123) [inline docs/Simulation/Software-RTL-Simulation.rst:135]: Accidental regression

File: sims/vcs/Makefile [hunk 2]
```diff
@@ -60,9 +138,26 @@ $(sim_debug): $(sim_vsrcs) $(sim_common_files) $(dramsim_lib) $(EXTRA_SIM_REQS)
 $(output_dir)/%.vpd: $(output_dir)/% $(sim_debug)
 	(set -o pipefail && $(sim_debug) $(PERMISSIVE_ON) $(SIM_FLAGS) $(EXTRA_SIM_FLAGS) $(VERBOSE_FLAGS) +vcdplusfile=$@ $(PERMISSIVE_OFF) $< </dev/null 2> >(spike-dasm > $<.out) | tee $<.log)
 
+$(output_dir)/none.vpd: $(sim_debug)
```
Comment (jerryz123) [inline sims/vcs/Makefile:141]: The run-none rules were removed.

File: sims/vcs/Makefile [hunk 3]
```diff
@@ -36,22 +39,97 @@ debug: $(sim_debug)
 #########################################################################################
 include $(base_dir)/common.mk
 
+#########################################################################################
+# verilator-specific user-interface variables and commands
+#########################################################################################
+HELP_COMPILATION_VARIABLES +=
```
Comment (jerryz123) [inline sims/vcs/Makefile:45]: Should this be duplicated across VCS/Verilator?

File: docs/Simulation/Software-RTL-Simulation.rst (hunk 1 above)
Comment (abejgonzalez) [reply docs/Simulation/Software-RTL-Simulation.rst:135]: Fixed.

File: sims/vcs/Makefile (hunk 3 above)
Comment (abejgonzalez) [reply sims/vcs/Makefile:42]: I moved it to `variables.mk` to reduce duplication.

File: sims/vcs/Makefile [hunk 4]
```diff
@@ -36,22 +36,97 @@ debug: $(sim_debug)
...
+
 #########################################################################################
 # vcs binary and arguments
 #########################################################################################
 VCS = vcs -full64
 
-VCS_OPTS = -notice -line $(VCS_CC_OPTS) $(VCS_NONCC_OPTS) $(VCS_DEFINE_OPTS) $(EXTRA_SIM_SOURCES)
+PREPROC_DEFINES = \
+	+define+VCS \
+	+define+CLOCK_PERIOD=1.0 \
```
Comment (jerryz123) [inline sims/vcs/Makefile:57]: VCS args were pulled into `chipyard/vcs.mk` since they are reused in hammer-sim. @colinschmidt

File: sims/vcs/Makefile [hunk 5]
```diff
@@ -36,22 +36,97 @@ debug: $(sim_debug)
...
 $(sim): $(sim_vsrcs) $(sim_common_files) $(dramsim_lib) $(EXTRA_SIM_REQS)
-	rm -rf csrc && $(VCS) $(VCS_OPTS) -o $@
+	rm -rf csrc && $(VCS) $(VCS_OPTS) $(EXTRA_SIM_SOURCES) -o $@ \
+	-debug_pp
 
 $(sim_debug): $(sim_vsrcs) $(sim_common_files) $(dramsim_lib) $(EXTRA_SIM_REQS)
-	rm -rf csrc && $(VCS) $(VCS_OPTS) -o $@ \
-	+define+DEBUG
+	rm -rf csrc && $(VCS) $(VCS_OPTS) $(EXTRA_SIM_SOURCES) -o $@ \
+	+define+DEBUG \
```
Comment (jerryz123) [inline sims/vcs/Makefile:128]: `-debug_pp` was removed due to duplication in `VCS_OPTS`

File: common.mk [hunk 6]
```diff
@@ -3,22 +3,46 @@
...
+"   EXTRA_GENERATOR_REQS   = requirements needed for the main generator" \
+"   EXTRA_SIM_CFLAGS       = CFLAGS for building simulators" \
+"   EXTRA_SIM_CXXFLAGS     = CXXFLAGS for building simulators" \
+"   EXTRA_SIM_LDFLAGS      = LDFLAGS for building simulators" \
+"   EXTRA_SIM_SOURCES      = simulation sources needed for simulator" \
+"   EXTRA_SIM_REQS         = requirements to build the simulator"
+
+EXTRA_GENERATOR_REQS ?=
+EXTRA_SIM_CXXFLAGS   ?=
+EXTRA_SIM_CFLAGS     ?=
```
Comment (jerryz123) [inline common.mk:25]: This variable appears to be unused. I don't think we need separate CXXFLAGS and CFLAGS

Comment (jerryz123) [reply common.mk:25]: Additionally, don't the makefiles which edit these variables (ariane,tracegen,nvdla,dromajo) need to be updated?

File: sims/vcs/Makefile (hunk 4 above)
Comment (abejgonzalez) [reply sims/vcs/Makefile:57]: This is the main reason why I marked this as "DO NOT MERGE". I have merged it now, but want to test VCS changes on the Millennium machines (might also make some Hammer changes).

File: common.mk (hunk 6 above)
Comment (abejgonzalez) [reply common.mk:25]: Correct, I remove the `CFLAGS` and just kept `CXXFLAGS`. The other `mk`s don't need to be changed since the names match (except for `dromajo.mk`).

File: sims/vcs/Makefile (hunk 5 above)
Comment (abejgonzalez) [reply sims/vcs/Makefile:128]: Yup. Already removed.

File: sims/vcs/Makefile (hunk 2 above)
Comment (abejgonzalez) [reply sims/vcs/Makefile]: Fixed.

File: vlsi/Makefile [hunk 7]
```diff
@@ -115,12 +115,16 @@ $(SIM_CONF): $(VLSI_RTL) $(HARNESS_FILE) $(HARNESS_SMEMS_FILE) $(sim_common_file
...
-	for x in $(filter-out "",$(filter-out -CC,$(VCS_CC_OPTS))); do \
+	echo "  compiler_cc_opts:" >> $@
+	for x in $(filter-out "",$(VCS_CXXFLAGS)); do \
+		echo '    - "'$$x'"' >> $@; \
+	done
+	echo "  compiler_ld_opts:" >> $@
+	for x in $(filter-out "",$(VCS_LDFLAGS)); do \
 		echo '    - "'$$x'"' >> $@; \
 	done
 	echo "  compiler_opts_meta: 'append'" >> $@
```
Comment (abejgonzalez) [inline vlsi/Makefile:130]: Do I need to change the `_meta` to apply to both the `_cc` and `_ld` keys?

Comment (colinschmidt) [reply vlsi/Makefile:130]: Yes. It does nothing the way it is now.

File: sims/verilator/Makefile [hunk 8]
```diff
@@ -47,67 +47,133 @@ debug: $(sim_debug)
...
+	--x-initial fast \
+	--assert \
+	--output-split 10000 \
+	--output-split-cfuncs 10000
+
+# default flags added for ariane (-Wno-fatal needed for -Wall to not cause
+# a crash, since 1000s of warnings are generated)
+VERILOG_IP_VERILATOR_FLAGS := \
 	--unroll-count 256 \
-	-Werror-PINMISSING \
```
Comment (colinschmidt) [inline sims/verilator/Makefile:77]: Whats the rational for removing these?

File: sims/verilator/Makefile [hunk 9]
```diff
@@ -47,67 +47,133 @@ debug: $(sim_debug)
...
+TRACING_OPTS := $(if $(filter $(VERILATOR_FST_MODE),0),\
+	                  --trace,--trace-fst --trace-threads 1)
+TRACING_CFLAGS := $(if $(filter $(VERILATOR_FST_MODE),0),,-DCY_FST_TRACE)
+
+#----------------------------------------------------------------------------------------
+# verilation configuration/optimization
+#----------------------------------------------------------------------------------------
+# we initially had --noassert for performance, but several modules use
+# assertions, including dramsim, so we enable --assert by default
+VMODE := \
```
Comment (colinschmidt) [inline sims/verilator/Makefile:83]: I don't like the name of this variable. It seems to be a kind of parent to the `IP_VERILATOR_FLAGS` and `CHIPYARD_VERILATOR_FLAGS` but its just called `VMODE` not `VERILATOR_FLAGS` or `COMMON_VERILATOR_FLAGS`.

File: sims/verilator/Makefile [hunk 10]
```diff
@@ -47,67 +47,133 @@ debug: $(sim_debug)
...
+	-Wno-fatal
+
+# normal flags used for chipyard builds (that are incompatible with vlog ip aka ariane)
+CHIPYARD_VERILATOR_FLAGS :=
 
-# normal flags used for chipyard builds (that are incompatible with ariane)
-CHIPYARD_VERILATOR_FLAGS = \
-	--assert
+# options dependent on whether ariane/NVDLA or chipyard is used
+# NOTE: defer the evaluation of this until it is used!
```
Comment (colinschmidt) [inline sims/verilator/Makefile:107]: I thought shell is evaluated before any rules are run. So I'm not sure how you are deferring evaluation.

File: sims/verilator/Makefile [hunk 11]
```diff
@@ -47,67 +47,133 @@ debug: $(sim_debug)
...
+#----------------------------------------------------------------------------------------
+# -flto slows down compilation on small-memory and breaks on firesim-manager
+CMODE := -O3 -fbranch-probabilities -march=native
+
+VERILATOR_CXXFLAGS = \
+	$(CXXFLAGS) \
+	$(RUNTIME_PROFILING_CFLAGS) \
+	$(TRACING_CFLAGS) \
+	$(CMODE) \
+	-std=c++11 \
```
Comment (colinschmidt) [inline sims/verilator/Makefile:146]: Why do all of these not get pulled into any of the above variables, or another variable?

File: sims/verilator/Makefile [hunk 12]
```diff
@@ -127,13 +193,13 @@ model_mk_debug = $(model_dir_debug)/V$(VLOG_MODEL).mk
 $(model_mk): $(sim_vsrcs) $(sim_common_files) $(EXTRA_SIM_REQS)
 	rm -rf $(model_dir)
 	mkdir -p $(model_dir)
-	$(VERILATOR) $(VERILATOR_OPTS) -o $(sim) -Mdir $(model_dir) -CFLAGS "-include $(model_header)"
+	$(VERILATOR) $(VERILATOR_OPTS) $(EXTRA_SIM_SOURCES) -o $(sim) -Mdir $(model_dir) -CFLAGS "-include $(model_header)"
```
Comment (colinschmidt) [inline sims/verilator/Makefile:196]: There are fancy ways(used in the vlsi directory) to DRY the flags that depend on whether you are building a debug simulator or not, but perhaps we should do that in a separate PR.

File: sims/verilator/Makefile [hunk 13]
```diff
@@ -47,67 +47,133 @@ debug: $(sim_debug)
 include $(base_dir)/common.mk
 
 #########################################################################################
-# verilator binary and flags
+# verilator-specific user-interface variables and commands
```
Comment (colinschmidt) [inline sims/verilator/Makefile:50]: I wonder if this should be pulled into a `verilator.mk` a la `vcs.mk`. I don't think we currently use it elsewhere but perhaps we will at some point, and in the meantime it adds some nice parallelism to the two currently supported simulators.

**Comment by abejgonzalez** [comment]
Ok the Chipyard simulation side of this should be good to review. I checked VCS simulation and Verilator simulations. Im trying to check Hammer sim right now.

File: variables.mk [hunk 14]
```diff
@@ -1,23 +1,51 @@
...
+HELP_PROJECT_VARIABLES = \
+"   SUB_PROJECT            = use the specific subproject default variables [$(SUB_PROJECT)]" \
+"   SBT_PROJECT            = the SBT project that you should find the classes/packages in [$(SBT_PROJECT)]" \
+"   MODEL                  = the top level module of the project in Chisel (normally the harness) [$(MODEL)]" \
+"   VLOG_MODEL             = the top level module of the project in Firrtl/Verilog (normally the harness) [$(VLOG_MODEL)]" \
+"   MODEL_PACKAGE          = the scala package to find the MODEL in [$(MODEL_PACKAGE)]" \
+"   CONFIG                 = the configuration class to give the parameters for the project [$(CONFIG)]" \
+"   CONFIG_PACKAGE         = the scala package to find the CONFIG class [$(CONFIG_PACKAGE)]" \
+"   GENERATOR_PACKAGE      = the scala package to find the Generator class in [$(GENERATOR_PACKAGE)]" \
+"   TB                     = wrapper over the TestHarness needed to simulate in a verilog simulator [$(TB)]" \
```
Comment (colinschmidt) [inline variables.mk:16]: I think we should have the words test bench in this description somewhere.

File: variables.mk [hunk 15]
```diff
@@ -1,23 +1,51 @@
...
+"   TOP                    = top level module of the project (normally the module instantiated by the harness) [$(TOP)]"
+
+HELP_SIMULATION_VARIABLES = \
+"   BINARY                 = riscv binary that the simulator will run" \
```
Comment (colinschmidt) [inline variables.mk:20]: `riscv elf binary that the simulator will run when using the run-binary* targets`

File: vcs.mk [hunk 16]
```diff
@@ -11,34 +11,56 @@ endif
 CLOCK_PERIOD ?= 1.0
 RESET_DELAY ?= 777.7
 
+#----------------------------------------------------------------------------------------
+# gcc configuration/optimization
+#----------------------------------------------------------------------------------------
+# -flto slows down compilation on small-memory and breaks on firesim-manager
+CMODE := -O3 -fbranch-probabilities -march=native
```
Comment (colinschmidt) [inline vcs.mk:18]: Where did this come from? Is it known to always be better than the default? Is it optimizing for compilation time or runtime? Is that what we want?

File: vcs.mk [hunk 17]
```diff
@@ -11,34 +11,56 @@ endif
...
+	-I$(RISCV)/include \
+	-I$(dramsim_dir) \
+	-std=c++11 \
+	$(EXTRA_SIM_CXXFLAGS)
+
+VCS_LDFLAGS =	\
+	$(LDFLAGS) \
+	-L$(RISCV)/lib \
+	-Wl,-rpath,$(RISCV)/lib \
+	-L$(sim_dir) \
```
Comment (colinschmidt) [inline vcs.mk:32]: There is getting to be an awful lot of overlap between these sets of flags for verilator and vcs, but I think we might want to wait for a different PR since we are already changing so much.

File: vcs.mk [hunk 18]
```diff
@@ -11,34 +11,56 @@ endif
...
+
+# vcs requires LDFLAGS to not include library names (i.e. -l needs to be separate)
 VCS_CC_OPTS = \
-	-CC "-I$(RISCV)/include" \
-	-CC "-I$(dramsim_dir)" \
-	-CC "-std=c++11" \
-	-CC "$(EXTRA_SIM_CC_FLAGS)"
+	-CFLAGS "$(VCS_CXXFLAGS)" \
+	-LDFLAGS "$(filter-out -l%,$(VCS_LDFLAGS))" \
+	$(filter -l%,$(VCS_LDFLAGS))
```
Comment (colinschmidt) [inline vcs.mk:42]: This makes me think the vlsi version of this is broken.

File: vcs.mk [hunk 19]
```diff
@@ -11,34 +11,56 @@ endif
...
 
 VCS_NONCC_OPTS = \
-	$(dramsim_lib) \
```
Comment (colinschmidt) [inline vcs.mk:21]: Where did these two flags go?

Comment (abejgonzalez) [reply vcs.mk:21]: They were moved into the LDFLAGs section (see the `-L` and `-l` flags)

File: vcs.mk (hunk 18 above)
Comment (abejgonzalez) [reply vcs.mk:42]: Yea I think I addressed it. Just haven't gotten around to checking the VLSI side yet.

File: vcs.mk (hunk 17 above)
Comment (abejgonzalez) [reply vcs.mk:32]: Yes. That was partly the intention of doing this... since if we can consolidate flags then this might be easier to deal with in the long run.

File: common.mk [hunk 20]
```diff
@@ -3,22 +3,44 @@
...
+
+EXTRA_GENERATOR_REQS ?=
+EXTRA_SIM_CXXFLAGS   ?=
+EXTRA_SIM_LDFLAGS    ?=
+EXTRA_SIM_SOURCES    ?=
+EXTRA_SIM_REQS       ?=
+
+#----------------------------------------------------------------------------
+HELP_SIMULATION_VARIABLES += \
+"   EXTRA_SIM_FLAGS        = runtime simulation flags (passed within +permissive)"
```
[the remark this answers is not in the document]
Comment (jerryz123) [reply common.mk:30]: Don't subproject Makefrags append to these?

File: vcs.mk (hunk 16 above)
Comment (abejgonzalez) [reply vcs.mk:18]: This was copying what Verilator did (though to make Verilator not segfault I just kept the `-O3`). It is optimizing the runtime. The default is nothing.

File: sims/verilator/Makefile (hunk 13 above)
Comment (jerryz123) [reply sims/verilator/Makefile:50]: I agree the parallelism would be nice. If we do that, I would like to see verilator.mk and vcs.mk moved into sims/verilator and sims/vcs

File: sims/verilator/Makefile [hunk 21]
```diff
@@ -155,8 +221,21 @@ $(output_dir)/%.vpd: $(output_dir)/% $(sim_debug)
...
+	rm -rf $(model_dir) $(sim)
+
+clean-sim-debug:
+	rm -rf $(model_dir_debug) $(sim_debug)
+
+#########################################################################################
+# print help text
+#########################################################################################
+.PHONY: help
+help:
```
Comment (jerryz123) [inline sims/verilator/Makefile:240]: Can this rule be deduplicated in `common.mk`

File: sims/verilator/Makefile (hunk 13 above)
Comment (colinschmidt) [reply sims/verilator/Makefile:50]: The reason `vcs.mk` is in the top-level is because its used by `vlsi`

File: vcs.mk (hunk 16 above)
Comment (colinschmidt) [reply vcs.mk:18]: Does it actually improve runtime?

File: vcs.mk (hunk 19 above)
Comment (colinschmidt) [reply vcs.mk:21]: Ah I didn't realize we wouldn't need the full path anymore, sorry.

File: sims/verilator/Makefile (hunk 13 above)
Comment (jerryz123) [reply sims/verilator/Makefile:50]: Yes, but `vlsi` can just look in `../sims/vcs` instead?

Comment (colinschmidt) [reply sims/verilator/Makefile:50]: Sure I guess that works as well. I don't have a strong preference. But I do feel like if we end up with a `sim.mk` `vcs.mk` and `verilator.mk` it would be slightly annoying to have them all in different directories.

File: common.mk [hunk 22]
```diff
@@ -3,22 +3,44 @@
...
+
+EXTRA_SIM_FLAGS ?=
+
+#----------------------------------------------------------------------------
+HELP_COMMANDS += \
+"   run-binary             = run [./$(shell basename $(sim))] and log instructions to file" \
+"   run-binary-fast        = run [./$(shell basename $(sim))] and don't log instructions" \
+"   run-binary-debug       = run [./$(shell basename $(sim_debug))] and log instructions and waveform to files" \
+"   verilog                = generate intermediate verilog files from chisel elaboration and firrtl passes"
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:39]: Up to us. Since this is a non-exhaustive list I don't think it will hurt to add one more

File: common.mk (hunk 20 above)
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:30]: I agree. You mean "additional runtime simulation flags"...?

File: common.mk [hunk 23]
```diff
@@ -3,22 +3,44 @@
...
 $(error RISCV is unset. You must set RISCV yourself, or through the Chipyard auto-generated env file)
 else
 $(info Running with RISCV=$(RISCV))
 endif
 
+#########################################################################################
+# specify user-interface variables
+#########################################################################################
+HELP_COMPILATION_VARIABLES += \
+"   EXTRA_GENERATOR_REQS   = requirements needed for the main generator" \
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:16]: Requirements are Make requirements.

File: common.mk [hunk 24]
```diff
@@ -3,22 +3,44 @@
...
+"   EXTRA_SIM_CXXFLAGS     = CXXFLAGS for building simulators" \
+"   EXTRA_SIM_LDFLAGS      = LDFLAGS for building simulators" \
+"   EXTRA_SIM_SOURCES      = simulation sources needed for simulator" \
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:19]: Correct.

File: common.mk [hunk 25]
```diff
@@ -3,22 +3,44 @@
...
+"   EXTRA_SIM_REQS         = requirements to build the simulator"
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:20]: Yup. Same answer as before just for building the simulator.

File: vcs.mk (hunk 16 above)
Comment (abejgonzalez) [reply vcs.mk:18]: I dunno. I can run the bmark-tests on Rocket with and without the `CMODE` flags. Im trying to get Hammer sim tested 1st.

Comment (abejgonzalez) [reply vcs.mk:18]: With CMODE flags:
Compile: 36s
Runtime for all bmark tests (serially): 7m 36s

Without CMODE flags:
Compile: 35s
Runtime for all bmark tests (serially): ???

File: sims/verilator/Makefile (hunk 21 above)
Comment (abejgonzalez) [reply sims/verilator/Makefile:240]: Yup

File: vcs.mk (hunk 16 above)
Comment (abejgonzalez) [reply vcs.mk:18]: While this doesn't result in a large runtime difference, I think it should stay if we end up consolidating flags across VCS and Verilator.

File: common.mk [hunk 26]
```diff
@@ -3,22 +3,44 @@
...
 #########################################################################################
-# extra make variables/rules from subprojects
-#
-# EXTRA_GENERATOR_REQS - requirements needed for the main generator
-# EXTRA_SIM_FLAGS - runtime simulation flags
-# EXTRA_SIM_CC_FLAGS - cc flags for simulators
-# EXTRA_SIM_SOURCES - simulation sources needed for simulator
-# EXTRA_SIM_REQS - requirements to build the simulator
+# include additional subproject make fragments
+# see HELP_COMPILATION_VARIABLES
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:44]: Printing out the values of these flags might be super verbose (especially if there is a lot of verilog files added). These aren't used often so Im not too concerned to add this.

File: common.mk [hunk 27]
```diff
@@ -139,19 +160,19 @@ verilog: $(sim_vsrcs)
 #########################################################################################
 # helper rules to run simulations
 #########################################################################################
-.PHONY: run-binary run-binary-fast run-binary-debug run-fast
+.PHONY: run-binary run-binary-fast
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply common.mk:164]: I think that was done by Sam in the past.

File: common.mk [hunk 28]
```diff
@@ -139,19 +161,19 @@ verilog: $(sim_vsrcs)
...
+.PHONY: run-binary-debug
+.PHONY: run-fast
```
Comment (abejgonzalez) [inline common.mk:166]:
```suggestion
.PHONY: run-binary run-binary-fast run-binary-debug run-fast
```

File: sims/vcs/Makefile [hunk 29]
```diff
@@ -61,8 +61,14 @@ $(output_dir)/%.vpd: $(output_dir)/% $(sim_debug)
...
 
 #########################################################################################
-# general cleanup rule
+# general cleanup rules
 #########################################################################################
-.PHONY: clean
+.PHONY: clean clean-sim clean-sim-debug
 clean:
-	rm -rf $(gen_dir) csrc $(sim_prefix)-* ucli.key vc_hdrs.h
+	rm -rf $(gen_dir) $(sim_prefix)-*
```
Comment (abejgonzalez) [inline sims/vcs/Makefile:68]:
```suggestion
	rm -rf $(gen_dir) csrc $(sim_prefix)-* ucli.key vc_hdrs.h
```

File: sims/verilator/Makefile (hunk 8 above)
Comment (abejgonzalez) [reply sims/verilator/Makefile:77]: These were removed by Sam in the past. I just copied this over.

File: sims/verilator/Makefile (hunk 10 above)
Comment (abejgonzalez) [reply sims/verilator/Makefile:107]: I remember that this was tricky to solve first. I had to use the `=` for it to work. Otherwise, this will search in the `build_dir` immediately and break.

File: sims/verilator/Makefile (hunk 12 above)
Comment (abejgonzalez) [reply sims/verilator/Makefile:196]: +1 to new PR's. I think this PR is getting larger by the min.

File: sims/verilator/Makefile (hunk 9 above)
Comment (abejgonzalez) [reply sims/verilator/Makefile:83]: I think this is to be consistent with `CMODE`. However, Im not opposed to changing both.

File: sims/verilator/Makefile [hunk 30]
```diff
@@ -47,67 +47,133 @@ debug: $(sim_debug)
...
+# verilator/cxx binary and flags
+#########################################################################################
+VERILATOR := verilator --cc --exe
 
-VERILATOR_CC_OPTS = \
+#----------------------------------------------------------------------------------------
+# user configs
+#----------------------------------------------------------------------------------------
+VERILATOR_PROFILE ?= none
+RUNTIME_PROFILING_CFLAGS := $(if $(filter $(VERILATOR_PROFILE),all),-g -pg,)
```
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply sims/verilator/Makefile:67]: Probably

File: sims/verilator/Makefile (hunk 11 above)
Comment (abejgonzalez) [reply sims/verilator/Makefile:146]: I think these are just general CFLAGS and thus they aren't profiling, tracing, or optimization flags...

File: sims/verilator/Makefile (hunk 30 above)
[the remark this answers is not in the document]
Comment (abejgonzalez) [reply sims/verilator/Makefile:67]: Actually, if you want this in one line then this is the good way to do it. Otherwise, you will have to do something like:

```
if (condition)
RUNTIME_PROFILING_CFLAGS := -g -pg
else
RUNTIME_PROFILING_CFLAGS :=
endif
```

**Comment by abejgonzalez** [comment]
Ok. I just got around to checking Hammer Sim. I got the simulator to compile successfully but I am getting other RISCV toolchains problems (causing wrong TSI commands to be sent). I think that is a problem with my environment though not the PR.

File: vcs.mk (hunk 18 above)
Comment (abejgonzalez) [reply vcs.mk:41]: This is now tested and should work properly. See the comment in the main PR thread.

File: sims/verilator/Makefile (hunk 10 above)
Comment (colinschmidt) [reply sims/verilator/Makefile:109]: Ok as long as you verified it adds the ariane flags even when there is an empty directory I'm satisfied.

File: sims/verilator/Makefile (hunk 12 above)
Comment (colinschmidt) [reply sims/verilator/Makefile:198]: It's clearly a fraught refactoring, given the previous incarnation had even more feedback. Thanks for working on it.

File: sims/verilator/Makefile (hunk 13 above)
Comment (colinschmidt) [reply sims/verilator/Makefile:50]: Will save this discussion for the PR that solves #653

File: vcs.mk (hunk 16 above)
Comment (colinschmidt) [reply vcs.mk]: Fine. It is worth noting that adding code that doesn't affect things is confusing, but if this proves that we can merge the flags in the future I'm fine with leaving it in.

**Comment by colinschmidt** [comment]
I filed issues for the things I suggested be updated in new PRs. I believe the HammerSim changes should work should I'm ok merging once CI passes. We're not close to a release now, so we have plenty of time to dog food the changes anyways.
